Frameworks/CoreGraphics/CGGradient.mm

```diff
@@ -18,186 +18,158 @@
 #import <StubReturn.h>
 #import <Starboard.h>
 #import <CoreGraphics/CGContext.h>
+#import <CFCppBase.h>
+#import <vector>
+#import <algorithm>
+#import <iterator>
+#import <LoggingNative.h>
 #import "CGColorSpaceInternal.h"
 #import "CGGradientInternal.h"
 #import "UIColorInternal.h"
-#import "_CGLifetimeBridgingType.h"
 
-@interface CGNSGradient : _CGLifetimeBridgingType
-@end
+static const wchar_t* TAG = L"CGGradient";
 
-@implementation CGNSGradient
-#pragma clang diagnostic push
-#pragma clang diagnostic ignored "-Wobjc-missing-super-calls"
-- (void)dealloc {
-    delete (__CGGradient*)self;
-}
-#pragma clang diagnostic pop
-@end
-
-__CGGradient::__CGGradient() : _components(NULL), _locations(NULL) {
-    object_setClass((id)this, [CGNSGradient class]);
-}
-
-__CGGradient::~__CGGradient() {
-    if (_components) {
-        delete[] _components;
-        _components = NULL;
-    }
-    if (_locations) {
-        delete[] _locations;
-        _locations = NULL;
+struct __CGGradient : CoreFoundation::CppBase<__CGGradient> {
+    __CGGradient(CGColorSpaceRef colorSpace = nullptr) : _colorSpace(colorSpace) {
     }
-}
 
-void __CGGradient::initWithColorComponents(const float* components, const float* locations, size_t count, CGColorSpaceRef colorspace) {
-    __CGColorSpace* cs = (__CGColorSpace*)colorspace;
-    DWORD componentCount = 0;
-
-    _colorSpaceModel = cs->colorSpaceModel;
-
-    switch (_colorSpaceModel) {
-        case kCGColorSpaceModelRGB:
-            componentCount = 4;
-            _format = _ColorABGR;
-            break;
-        case kCGColorSpaceModelMonochrome:
-            componentCount = 2;
-            _format = _ColorGrayscale;
-            break;
-        default:
-            UNIMPLEMENTED_WITH_MSG("Unsupported colorspace used to create gradiant.");
-            break;
+    __CGGradient(CGColorSpaceRef colorSpace, const CGFloat* components, const CGFloat* locations, size_t count) : __CGGradient(colorSpace) {
```

Comment:
You can use the vector's built-in iterator-range copy constructor for this! It'll let you simplify a lot.

I encourage you to think about how you can empty out all your constructor bodies and move things into _initializer lists_. :smile: